File: QMPrompter/Services/ScrollEngine.swift
import Combine
import QuartzCore
import SwiftUI

@MainActor
final class ScrollEngine: ObservableObject {
    @Published var offset: CGFloat = 0
    @Published var isPlaying = false
    @Published var speed: Double = 80

    private var displayLink: CADisplayLink?
    private var lastTimestamp: CFTimeInterval = 0
    private var lineHeight: CGFloat = 84
    private var averageCharactersPerLine: CGFloat = 18
    private var maximumOffset: CGFloat = 0
    private var followTargetOffset: CGFloat?

    deinit {
        displayLink?.invalidate()
    }

    func configure(
        speed: Double,
        lineHeight: CGFloat,
        averageCharactersPerLine: CGFloat,
        maximumOffset: CGFloat
    ) {
        self.speed = speed
        self.lineHeight = max(40, lineHeight)
        self.averageCharactersPerLine = max(6, averageCharactersPerLine)
        self.maximumOffset = max(0, maximumOffset)
        offset = min(offset, self.maximumOffset)
        if let followTargetOffset {
            self.followTargetOffset = min(self.maximumOffset, max(0, followTargetOffset))
        }
    }

    func play() {
        guard !isPlaying else { return }
        followTargetOffset = nil
        isPlaying = true
        ensureDisplayLink()
    }

    func pause() {
        isPlaying = false
    }

    func toggle() {
        isPlaying ? pause() : play()
    }

    func setSpeed(_ value: Double) {
        speed = min(260, max(20, value))
    }

    func setOffset(_ value: CGFloat) {
        followTargetOffset = nil
        applyOffset(value)
    }

    func follow(to value: CGFloat) {
        followTargetOffset = min(maximumOffset, max(0, value))
        isPlaying = false
        ensureDisplayLink()
    }

    func stopFollowing() {
        followTargetOffset = nil
    }

    private func applyOffset(_ value: CGFloat) {
        offset = min(maximumOffset, max(0, value))
        if offset >= maximumOffset {
            pause()
        }
    }

    func reset() {
        offset = 0
        followTargetOffset = nil
        pause()
    }

    private func ensureDisplayLink() {
        guard displayLink == nil else { return }
        let link = CADisplayLink(target: self, selector: #selector(tick(_:)))
        if #available(iOS 15.0, *) {
            link.preferredFrameRateRange = CAFrameRateRange(minimum: 30, maximum: 60, preferred: 60)
        } else {
            link.preferredFramesPerSecond = 60
        }
        link.add(to: .main, forMode: .common)
        displayLink = link
    }

    @objc private func tick(_ link: CADisplayLink) {
        guard isPlaying || followTargetOffset != nil else {
            lastTimestamp = link.timestamp
            return
        }

        if lastTimestamp == 0 {
            lastTimestamp = link.timestamp
            return
        }

        let delta = link.timestamp - lastTimestamp
        lastTimestamp = link.timestamp

        if let target = followTargetOffset {
            let response = min(1, CGFloat(delta) * 12)
            let nextOffset = offset + (target - offset) * response

            if abs(nextOffset - target) < 0.5 {
                applyOffset(target)
                followTargetOffset = nil
            } else {
                applyOffset(nextOffset)
            }
            return
        }

        let visualTuningFactor = 1.85
        let linesPerSecond = (speed / Double(averageCharactersPerLine)) / 60 * visualTuningFactor
        let pixelsPerSecond = CGFloat(linesPerSecond) * lineHeight
        applyOffset(offset + pixelsPerSecond * CGFloat(delta))
    }
}

File: app/src/main/kotlin/com/qiaomu/prompter/ui/prompter/ScrollEngine.kt
package com.qiaomu.prompter.ui.prompter

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableDoubleStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.withFrameNanos
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.isActive
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

class ScrollEngine(
    initialSpeed: Double = 80.0
) {
    var offset by mutableDoubleStateOf(0.0)
        private set

    var isPlaying by mutableStateOf(false)
        private set

    var speed by mutableDoubleStateOf(initialSpeed.coerceIn(MIN_SPEED, MAX_SPEED))
        private set

    private var lineHeight = 84.0
    private var averageCharactersPerLine = 18.0
    private var maximumOffset = 0.0
    private var followTargetOffset: Double? = null

    fun configure(
        speed: Double,
        lineHeight: Double,
        averageCharactersPerLine: Double,
        maximumOffset: Double
    ) {
        this.speed = speed.coerceIn(MIN_SPEED, MAX_SPEED)
        this.lineHeight = max(40.0, lineHeight)
        this.averageCharactersPerLine = max(6.0, averageCharactersPerLine)
        this.maximumOffset = max(0.0, maximumOffset)
        offset = offset.coerceIn(0.0, this.maximumOffset)
        followTargetOffset = followTargetOffset?.coerceIn(0.0, this.maximumOffset)
    }

    fun play() {
        followTargetOffset = null
        isPlaying = true
    }

    fun pause() {
        isPlaying = false
    }

    fun toggle() {
        if (isPlaying) pause() else play()
    }

    fun updateSpeed(value: Double) {
        speed = value.coerceIn(MIN_SPEED, MAX_SPEED)
    }

    fun updateOffset(value: Double) {
        followTargetOffset = null
        applyOffset(value)
    }

    fun follow(to: Double) {
        followTargetOffset = to.coerceIn(0.0, maximumOffset)
        isPlaying = false
    }

    fun stopFollowing() {
        followTargetOffset = null
    }

    fun reset() {
        offset = 0.0
        followTargetOffset = null
        pause()
    }

    suspend fun runFrameLoop() {
        var lastFrameNanos = 0L
        while (currentCoroutineContext().isActive) {
            val frameNanos = withFrameNanos { it }
            if (!isPlaying && followTargetOffset == null) {
                lastFrameNanos = frameNanos
                continue
            }
            if (lastFrameNanos == 0L) {
                lastFrameNanos = frameNanos
                continue
            }

            val delta = (frameNanos - lastFrameNanos) / 1_000_000_000.0
            lastFrameNanos = frameNanos

            val target = followTargetOffset
            if (target != null) {
                val response = min(1.0, delta * 12.0)
                val nextOffset = offset + (target - offset) * response
                if (abs(nextOffset - target) < 0.5) {
                    applyOffset(target)
                    followTargetOffset = null
                } else {
                    applyOffset(nextOffset)
                }
            } else {
                val linesPerSecond = (speed / averageCharactersPerLine) / 60.0 * VISUAL_TUNING_FACTOR
                val pixelsPerSecond = linesPerSecond * lineHeight
                applyOffset(offset + pixelsPerSecond * delta)
            }
        }
    }

    private fun applyOffset(value: Double) {
        offset = value.coerceIn(0.0, maximumOffset)
        if (offset >= maximumOffset) {
            pause()
        }
    }

    companion object {
        private const val MIN_SPEED = 20.0
        private const val MAX_SPEED = 260.0
        private const val VISUAL_TUNING_FACTOR = 1.85
    }
}
